### PR TITLE
[ticket/16370] Add core events for Notification

### PR DESCRIPTION
**Pull request by dark-1**
This core events allow extensions to control & manipulate Notifications.

Checklist:

- [x] Correct branch: master for new features; 3.3.x & 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html), [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-16370

### CODE REVIEW

**Comment by dark-1** [comment]
!set Event

**Comment by dark-1** [comment]
@mrgoldy ,
some review on this too , Please...

Best regards 👍

**Comment by dark-1** [comment]
@marc1706 ,

May I know the reason for this PR is on hold ?

Thanking you , best regards 👍

**Comment by DavidIQ** [comment]
What gives you the idea that this is "on hold"?

**Comment by dark-1** [comment]
> What gives you the idea that this is "on hold"?

I didn't mean to offend ,
But the PRs which I had done later than this PR were just merge less than an hour ago...
Hence I assumed that this PR is on hold.

Best of regards 👍

**Comment by marc1706** [comment]
Just means that I don't review PRs via a FIFO queue. ;)

**Comment by dark-1** [comment]
> Just means that I don't review PRs via a FIFO queue. ;)

Hahaha 😁
Then how is it ?

Also I had a proposal ,
The functions `output_notification_types` & `output_notification_methods` which are having global variables as function parameters,
to be removed & initialize them as global variables in the functions body.

Should I include this changes in this PR or make a new PR?
If this changes are applied, then will there be any issues?

Best of regards 👍

File: phpBB/includes/ucp/ucp_notifications.php [hunk 1]
```diff
@@ -57,15 +57,34 @@ public function main($id, $mode)
 
 					foreach ($phpbb_notifications->get_subscription_types() as $group => $subscription_types)
 					{
-						foreach ($subscription_types as $type => $data)
+						foreach ($subscription_types as $type => $type_data)
 						{
 							foreach ($notification_methods as $method => $method_data)
 							{
-								if ($request->is_set_post(str_replace('.', '_', $type . '_' . $method_data['id'])) && (!isset($subscriptions[$type]) || !in_array($method_data['id'], $subscriptions[$type])))
+								$is_set_notify = ($request->is_set_post(str_replace('.', '_', $type . '_' . $method_data['id']))) ? true : false;
```
Comment (marc1706) [inline phpBB/includes/ucp/ucp_notifications.php]: Can be simplified:
```suggestion
								$is_set_notify = $request->is_set_post(str_replace('.', '_', $type . '_' . $method_data['id']));
```

File: phpBB/includes/ucp/ucp_notifications.php [hunk 2]
```diff
@@ -168,9 +187,12 @@ public function main($id, $mode)
 	* @param \phpbb\notification\manager $phpbb_notifications
 	* @param \phpbb\template\template $template
 	* @param \phpbb\user $user
+	* @param \phpbb\event\dispatcher_interface $phpbb_dispatcher
 	* @param string $block
+	*
+	* @changed 3.2.10-RC1 Added $phpbb_dispatcher
```
Comment (marc1706) [inline phpBB/includes/ucp/ucp_notifications.php]: `@changed` shouldn't be used for method docblocks.

File: phpBB/phpbb/notification/manager.php [hunk 3]
```diff
@@ -333,10 +333,34 @@ public function add_notifications_for_users($notification_type_name, $data, $not
...
 			}
+
+			/**
+			* Allow filtering the notify_users array by notification_type_name & notification_method_name for a notification that is about to be sent.
+			* Here, $notify_users is already filtered from users who've already been notified.
+			*
+			* @event core.notification_manager_add_notifications_for_users_modify_data
+			* @var	string	notification_type_name		The notification type identifier
+			* @var	string	notification_method_name	The notification method identifier (read only)
+			* @var	array 	data						Data specific for the notification_type_name used will be inserted
```
Comment (marc1706) [inline phpBB/phpbb/notification/manager.php]: The docblock description above made more sense: `Data specific for this type that will be inserted`

File: phpBB/phpbb/notification/manager.php [hunk 4]
```diff
@@ -333,10 +333,34 @@ public function add_notifications_for_users($notification_type_name, $data, $not
...
+			* @var	array 	notify_users				The array of userid that are going to be notified for this notification. Set to array() to cancel or unset each user by userid.
```
Comment (marc1706) [inline phpBB/phpbb/notification/manager.php]: Same as above. I think the original description from the docblock is clear enough and more on point.

Comment (dark-1) [reply phpBB/phpbb/notification/manager.php]: Can you please clarify this , what you want me to do here ???
I copied from https://github.com/phpbb/phpbb/pull/5862/files#diff-0d7ecc527ae7eeebc4f109c2fdfcb1abR282

Best regards 👍

**Comment by dark-1** [comment]
@marc1706 ,

Did you go through this comment ? : https://github.com/phpbb/phpbb/pull/5862#issuecomment-594076144
If not then kindly give a look, please.

Should I do those changes ?

Best regards 👍

**Comment by marc1706** [comment]
Those were specifically moved to no longer use the `global` keyword. This should not be reinstated as we try to move away from using `global` where possible.

**Comment by dark-1** [comment]
> Those were specifically moved to no longer use the `global` keyword. This should not be reinstated as we try to move away from using `global` where possible.

Ahh!! ok .

Thanks for the info. 😄 

Best regards 👍

**Pull request merged**